### ISSUE TITLE
DaemonSet should not schedule pods on NotReady nodes

### ISSUE THREAD

**Issue by paralin** [kubernetes/kubernetes#16105]
It appears in my cluster that DaemonSet is scheduling pods on NotReady nodes.

To reproduce:

  1. Spin up a 6 node cluster on AWS
  2. Create a DaemonSet
  3. Scale back down to 3 nodes on AWS
  4. DaemonSet will continue to schedule and reschedule pods on the NotReady nodes.

Sample node list output:

```
NAME                                         LABELS                                                              STATUS     AGE
ip-172-20-0-168.us-west-2.compute.internal   kubernetes.io/hostname=ip-172-20-0-168.us-west-2.compute.internal   NotReady   13h
ip-172-20-0-169.us-west-2.compute.internal   kubernetes.io/hostname=ip-172-20-0-169.us-west-2.compute.internal   NotReady   13h
ip-172-20-0-170.us-west-2.compute.internal   kubernetes.io/hostname=ip-172-20-0-170.us-west-2.compute.internal   NotReady   13h
ip-172-20-0-244.us-west-2.compute.internal   kubernetes.io/hostname=ip-172-20-0-244.us-west-2.compute.internal   Ready      12m
ip-172-20-0-245.us-west-2.compute.internal   kubernetes.io/hostname=ip-172-20-0-245.us-west-2.compute.internal   Ready      12m
ip-172-20-0-246.us-west-2.compute.internal   kubernetes.io/hostname=ip-172-20-0-246.us-west-2.compute.internal   Ready      12m
```

Sample pods output:

```
NAME              READY     STATUS        RESTARTS   AGE
cassandra-joag3   0/1       Terminating   0          47s
cassandra-jyjab   0/1       Terminating   0          57s
cassandra-l9o8b   1/1       Running       0          12m
cassandra-qmcw2   0/1       Terminating   0          37s
cassandra-r31tf   1/1       Running       0          12m
cassandra-ua95o   1/1       Running       0          12m
```

Sample DaemonSet output:

```
NAME        CONTAINER(S)   IMAGE(S)                                SELECTOR         NODE-SELECTOR
cassandra   cassandra      gcr.io/google_containers/cassandra:v6   name=cassandra   <none>
```

**Comment by paralin** [comment]
Confirming this is a pretty major bug. It still seems to even try to schedule pods on nodes after I delete them.

Pods seem to also be stuck in the "Terminating" state even when I manually delete them. What gives?

**Comment by davidopp** [comment]
@erictune Can you look into this?

**Comment by davidopp** [comment]
@mikedanese said he'd take a look.

I'll leave this in 1.1 milestone until he renders a verdict.

**Comment by mikedanese** [comment]
I thing we _want_ daemonset to place daemon pods onto nodes that are not ready. See #1574. Daemon pods should have the highest level of forgiveness.

**Comment by paralin** [comment]
@mikedanese If you do this, they instantly terminate. So the pods are constantly in a loop of being created -> terminate -> create -> terminate... doesn't make much sense.

**Comment by davidopp** [comment]
I think this is debatable -- definitely we want daemon pods to keep running on nodes when they become NotReady, but it's not obvious to me that we want to schedule new pods onto nodes that are NotReady.

**Comment by davidopp** [comment]
@bgrant0607

**Comment by bgrant0607** [comment]
We want daemons on nodes before anything else can schedule, and we want daemons to stay on nodes that become NotReady. 

In general, we need to evolve the node lifecycle more. We don't currently have a handshake that would enable daemons to be placed before anything else. Scheduling on NotReady nodes is one way to make that happen. In any case, Kubelet and DaemonSet should handle this gracefully.

This isn't P0. DaemonSet is alpha.

**Comment by mikedanese** [comment]
@paralin I would argue that that bug is better fixed by making nodecontroller not evict daemon pods, not by making daemonset not place pods on NotReady nodes.

**Comment by mikedanese** [comment]
Which is part of #15310

**Comment by davidopp** [comment]
>  We don't currently have a handshake that would enable daemons to be placed before anything else.  Scheduling on NotReady nodes is one way to make that happen.

I guess, but I think you'd agree this is a pretty bad way to do it.

Anyway, I'll remove this from 1.1

**Comment by mikedanese** [comment]
> DaemonSet should not schedule pods on NotReady nodes

We can do this for v1.1 to fix the create delete loop, and implement the real solution after?

**Comment by davidopp** [comment]
> We can do this for v1.1 to fix the create delete loop, and implement the real solution after?

Yeah I guess the question is, for 1.1, which is worse: the create-delete loop, or having a daemon be unable to schedule onto the node because other pods got there first.

Since the second thing can happen even with the current behavior (for example, the node transitions to Ready and gets regular pods scheduled onto it before the DaemonSet controller schedules the daemons), I'd favor doing what the OP suggested (DaemonSet should not schedule pods on NotReady nodes) for 1.1, and doing the more proper solution(s) Brian alluded to for 1.2.

**Comment by mikedanese** [comment]
fix in #16478

**Comment by erictune** [comment]
Someone 
9maybe @mikedanese) should create a new issue to do the behavior @bgrant0607 mentioned before DaemonSet goes beta, and then mention that issue as a graduation-to-beta requirement on #15310, and then close this issue.

**Comment by mikedanese** [comment]
> Either integration into the node controller in order to integrate with node lifecycle, such as creation before other pods can be scheduled and implicit forgiveness behavior (#1574), or use of initializers and finalizers (#3585) to achieve the same

is already part of #15310 but is a gradutation for v1, no beta. Should that be bumped to beta graduation requirements? Anything that isn't covered by the bullet?

**Comment by erictune** [comment]
Ok, fine as is.

**Comment by bgrant0607** [comment]
Waiting on merge and cherrypick
